apps/src/lab2/projects/SourcesStore.ts

```diff
@@ -59,7 +69,13 @@ export class RemoteSourcesStore implements SourcesStore {
       this.firstSaveTime = this.firstSaveTime || timestamp;
       this.currentVersionId = versionId;
     } else {
-      throw new Error(response.status + ' ' + response.statusText);
+      const reader = await response.json();
+      const details = reader.details;
+      const statusText =
+        details && details.profaneWords
+          ? `${response.statusText} (profanity)`
```

Comment:
Hmm, this got me thinking about ways to avoid having to inspect the response and append a specific string here in the Lab2 layer. For example if we have other specific error types that different labs want to look out for, we could end up with a lot of special cases here. This would be out of scope for this PR, but one thing we could do is update `sourcesApi` to use the `HttpClient` put method, which will throw a `NetworkError` if a request fails (see [here](https://github.com/code-dot-org/code-dot-org/blob/d971cd2ff302dc6088c253cd071ae3911e3b83ed/apps/src/util/HttpClient.ts#L19)). That `NetworkError` type stores a reference to the response object, which then would be passed to AI Chat in the onSaveFail error handler. AI Chat could then inspect the response and decide how to handle it appropriately. 

For now, another option could be to just change the `Error` thrown here to a `NetworkError` and pass the response (i.e. `throw new NetworkError(response.status + ' ' + statusText, response);`). Then AI Chat can read the body JSON and show the appropriate error message.

Reply:
Thanks! I'll update `Error` to `Network` error - appreciate the suggestion!